presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java

```diff
@@ -76,42 +88,57 @@ public Parts createPartialAggregations(PlanVariableAllocator variableAllocator,
                     originalAggregation.getOrderBy(),
                     originalAggregation.isDistinct(),
                     originalAggregation.getMask()));
-            finalAggregation.put(entry.getKey(),
+
+            // create final aggregation
+            finalOrIntermediateAggregations.put(entry.getKey(),
                     new Aggregation(
                             new CallExpression(
                                     originalAggregation.getCall().getDisplayName(),
                                     functionHandle,
-                                    function.getFinalType(),
+                                    intermediate ? function.getIntermediateType() : function.getFinalType(),
                                     ImmutableList.of(partialVariable)),
                             Optional.empty(),
                             Optional.empty(),
                             false,
                             Optional.empty()));
         }
+
+        StatisticAggregations finalOrIntermediateAggregation = new StatisticAggregations(finalOrIntermediateAggregations.build(), groupingVariables);
         return new Parts(
-                new StatisticAggregations(partialAggregation.build(), groupingVariables),
-                new StatisticAggregations(finalAggregation.build(), groupingVariables));
+                intermediate ? Optional.empty() : Optional.of(finalOrIntermediateAggregation),
+                intermediate ? Optional.of(finalOrIntermediateAggregation) : Optional.empty(),
+                new StatisticAggregations(partialAggregations.build(), groupingVariables));
     }
 
     public static class Parts
     {
+        private final Optional<StatisticAggregations> finalAggregation;
+        private final Optional<StatisticAggregations> intermediateAggregation;
         private final StatisticAggregations partialAggregation;
-        private final StatisticAggregations finalAggregation;
 
-        public Parts(StatisticAggregations partialAggregation, StatisticAggregations finalAggregation)
+        public Parts(
+                Optional<StatisticAggregations> finalAggregation,
+                Optional<StatisticAggregations> intermediateAggregation,
+                StatisticAggregations partialAggregation)
         {
-            this.partialAggregation = requireNonNull(partialAggregation, "partialAggregation is null");
             this.finalAggregation = requireNonNull(finalAggregation, "finalAggregation is null");
+            this.intermediateAggregation = requireNonNull(intermediateAggregation, "intermediateAggregation is null");
```

Comment:
Maybe check one and only one of `intermediateAggregation` and `finalAggregation` is presented?